### PR TITLE
add simple cnn and rnn pytorch models

### PR DESCRIPTION
**Pull request by erobitschek**
refactored the code and expanded model repertoire

### CODE REVIEW

**Comment by Simsso** [comment]
Did you get a chance to test these new models already on (toy) data?

File: src/configs/config_scaffold.py [hunk 1]
```diff
@@ -46,16 +74,46 @@ def __post_init__(self):
...
+    batch_size: int = 32
     epochs: int = 500
+    model_type: ModelType = ModelType.LOGREG_SKLEARN
     framework: ModelFramework = ModelFramework.SKLEARN
+    loss_criterion: InitVar[
+        Optional[str]
+    ] = "CROSS_ENTROPY"  # default value for pytorch models
+    optimizer: InitVar[Optional[str]] = "ADAM"  # default value for pytorch models
+    actual_loss_criterion: Optional[Type[nn.Module]] = field(default=None, init=False)
+    actual_optimizer: Optional[Type[optim.Optimizer]] = field(default=None, init=False)
```
[the remark this answers is not in the document]
Comment (Simsso) [reply src/configs/config_scaffold.py:86]: If you initialize the fields with `None`, just put `= None` there.  No need for `field` afaik.

File: src/configs/config_scaffold.py [hunk 2]
```diff
@@ -46,16 +74,46 @@ def __post_init__(self):
...
+    data_transforms: list = field(
+        default_factory=lambda: None
+    )  # list of transforms to apply to data
     dropout_rate: float = 0.5
     patience: int = None
     params: dict = None  # param dict for lgbm model
     param_grid: dict = None  # for grid search to find optimal model parameters
     grid_search: bool = False  # whether to perform the grid search
 
-    # TODO: add post init that adds some parameters based on framework
```
Comment (Simsso) [inline src/configs/config_scaffold.py:58]: Nice when TODOs get deleted :-)

File: src/configs/config_scaffold.py [hunk 3]
```diff
@@ -46,16 +74,46 @@ def __post_init__(self):
...
+    def __post_init__(self, loss_criterion: str, optimizer: str):
+        if self.framework == ModelFramework.PYTORCH:
+            self.set_pytorch_components(loss_criterion, optimizer)
+        else:
+            # TODO: Implement logic for other frameworks if needed
+            pass
```
Comment (Simsso) [inline src/configs/config_scaffold.py:101]: You can replace TODO and pass by `raise NotImplementedError("Framework not supported")`

File: src/configs/config_scaffold.py [hunk 4]
```diff
@@ -46,16 +74,46 @@ def __post_init__(self):
...
+            if loss_criterion
+            else PyTorchLoss.CROSS_ENTROPY.value
+        )
+        optimizer_class = (
+            PyTorchOptim[optimizer].value if optimizer else PyTorchOptim.ADAM.value
+        )
+
+        object.__setattr__(
+            self, "actual_loss_criterion", loss_class
+        )  # this logic is needed to avoid dataclass immutability
```
Comment (Simsso) [inline src/configs/config_scaffold.py:115]: If you need to modify attributes afterwards, you can just remove `frozen=True`. Better for a reader who would then know that fields may change.\

File: src/configs/models.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        """
+        Parameters
+        ----------
+        input_dim: shape of input (batch, input_dims, sequence length)
+        n_class: number of classes to be categorized
+        filter_sizes: filter depth for Conv1d layers
+        kernel_sizes: kernel sizes for Conv1d layers
+        pooling_sizes: pooling window sizes
+        dense_sizes: size of each output sample for linear layers
+        """
```
Comment (Simsso) [inline src/configs/models.py:86]: Different docstring notation suddenly. Maybe a docstring isn't needed here.

File: src/configs/models.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        # Define convolutional layers
+        conv1d_layers = []
+        for in_ch, out_ch, kernel_size in zip(
+            [
+                self.input_dim[1],
+            ]
+            + self.filter_sizes[:-1],
+            self.filter_sizes,
+            self.kernel_sizes,
+        ):
```
[the remark this answers is not in the document]
Comment (Simsso) [reply src/configs/models.py:108]: You don't need the `self.` for all variables. Perhaps it fits into one line and is more readable then. If not, you can pull out the first of the three lists and construct it before the loop.

File: src/configs/models.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+            conv1d_layers.append(nn.Conv1d(in_ch, out_ch, kernel_size, padding=0))
+        self.conv1d_layers = nn.ModuleList(conv1d_layers)
+
+        # Define pooling layers
```
Comment (Simsso) [inline src/configs/models.py:112]: Unnecessary comment. If you think this block needs a comment, put it into a separate function `self._define_pooling_layers()` so the function name documents what's going on.

Comment (Simsso) [reply src/configs/models.py:112]: Same for the other code blocks of that kind then.

File: src/configs/models.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+
+        # Define linear layers (assumes that the first linear layer follows last conv1d layer)
+        linear_layers = []
+        for in_feat, out_feat in zip(
+            [
+                self.filter_sizes[-1],
+            ]
+            + self.dense_sizes[:-1],
+            self.dense_sizes,
+        ):
```
Comment (Simsso) [inline src/configs/models.py:129]: Big zip, hard to read.

File: src/configs/models.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+            linear_layers.append(nn.Linear(in_feat, out_feat))
+
+        self.linear_layers = nn.ModuleList(linear_layers)
+
+        # Define output layer
+        self.output_layer = nn.Linear(self.dense_sizes[-1], self.n_class)
+
+        # Activation function
```
Comment (Simsso) [inline src/configs/models.py:137]: Unnecessary

File: src/configs/models.py [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+        self.relu = nn.ReLU()
+
+    def forward(self, x: torch.Tensor):
+        """Defines the forward pass for the SimpleCNN.
+
+        Args:
+            x (torch.Tensor): Input tensor.
+
+        Returns:
+            torch.Tensor: Network output tensor.
```
Comment (Simsso) [inline src/configs/models.py:147]: These docstrings are useless, unless they specify of input and output. That's what a caller wants to know, I'd say.

File: src/configs/models.py [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+        # Transition from convolutional to linear layers.
+        x = self.adaptive_max_pool(x)
+        x = torch.squeeze(x, dim=-1)
+
+        # Process through linear layers.
+        for c_layer in self.linear_layers:
+            x = c_layer(x)
+            x = self.relu(x)
+
+        # Process through output layer and apply softmax
```
Comment (Simsso) [inline src/configs/models.py:164]: Where is the softmax being applied?

File: src/configs/models.py [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+        x = self.output_layer(x)
+
+        # Return network response
```
Comment (Simsso) [inline src/configs/models.py:167]: Unnecessary

File: src/configs/models.py (hunk 11 above)
Comment (Simsso) [reply src/configs/models.py:164]: I think you want to return the logits here. (So the comment is wrong, but the code correct.)

File: src/configs/models.py [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+            "n_classes": self.n_class,
+            "filter_sizes": self.filter_sizes,
+            "kernel_sizes": self.kernel_sizes,
+            "pooling_sizes": self.pooling_sizes,
+            "dense_sizes": self.dense_sizes,
+        }
+
+
+class SimpleRNN(nn.Module):
+    """Implements a basic RNN model using PyTorch.
```
Comment (Simsso) [inline src/configs/models.py:189]: Every class "implements" so that can be removed.

Comment (Simsso) [reply src/configs/models.py:189]: Maybe what is "basic" about it is worth expanding instead.

Comment (Simsso) [reply src/configs/models.py:189]: Or put differently: "implements" adds zero info and "basic" makes me ask questions. Instead the questions that "basic" raises should be answered and "basic" can then be removed.

File: src/configs/models.py [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+        sequence_length: Length of the sequence.
+        n_classes: Number of classes.
+        rnn: The core RNN module.
+        flatten: Flatten layer.
+        fc: Fully connected layer for output.
+    """
+
+    def __init__(
+        self,
+        input_dim: npt.ArrayLike,
```
Comment (Simsso) [inline src/configs/models.py:204]: Why not a list of ints here?

Comment (Simsso) [reply src/configs/models.py:204]: Maybe "shape" is better than "dim".

File: src/configs/models.py [hunk 15]
```diff
@@ -0,0 +1,283 @@
+from enum import Enum
+from typing import Callable, List, Optional, Tuple
+
+import numpy.typing as npt
+import torch as torch
+import torch.nn as nn
+from lightgbm import LGBMClassifier
+from sklearn.linear_model import LogisticRegression as skLogisticRegression
```
Comment (Simsso) [inline src/configs/models.py:8]: Keep class naming convention when renaming upon importing. Also why the rename? (It's fine, but be sure to have a reason.)

File: src/configs/models.py [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+        }
+
+
+class ModelType(Enum):
+    LOGREG_SKLEARN = skLogisticRegression
+    LOGREG_TORCH = TorchLogisticRegression
+    LGBM_CLASSIFIER = LGBMClassifier
+    SIMPLE_CNN = SimpleCNN
+    SIMPLE_RNN = SimpleRNN
+    # Add other models here as needed
```
Comment (Simsso) [inline src/configs/models.py:272]: No need to say that. It's clear that an enum can and should be expanded.

File: src/configs/models.py [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+
+    def __call__(self, *args, **kwargs) -> Callable:
+        # This allows the enum instance itself to be called as a constructor, forwarding arguments to the actual constructor
```
Comment (Simsso) [inline src/configs/models.py:275]: Neat

File: src/data.py [hunk 18]
```diff
@@ -51,16 +51,18 @@ def __getitem__(self, idx):
...
 
 
-def load_data(path: str, filter_cols: list = None) -> pd.DataFrame:
+def load_data(path: str, filter_cols: list = None, header: bool = True) -> pd.DataFrame:
     """Loads data from a CSV file and optionally filters certain columns."""
-    data = pd.read_csv(path)
+    if header:
+        data = pd.read_csv(path)
+    else:
+        data = pd.read_csv(path, header=None)
```
Comment (Simsso) [inline src/data.py:65]: Make `header` a string with default value `"infer"` and save yourself the if/else here.

File: src/data.py [hunk 19]
```diff
@@ -341,3 +356,39 @@ def prep_data_for_modelling(
...
+
+
+def prep_data_for_modelling(
+    config: RunConfig,
+    run_dir: str,
+    data_state: str,
+    logger: logging.Logger,
+    to_array: bool = True,
+) -> tuple[DataSplit, DataSplit, Optional[DataSplit]]:
+    """Prepares data for modeling using the given configuration.
```
Comment (Simsso) [inline src/data.py:368]: Inconsistent spelling of `modeling`.

File: src/data.py [hunk 20]
```diff
@@ -341,3 +356,39 @@ def prep_data_for_modelling(
...
+            config=config, logger=logger, val_frac=0.15
+        )
+    elif config.dataset.project == "synth_med_data":
+        train_set, test_set, val_set = prep_synth_med_data_for_modelling(
+            config=config,
+            run_dir=run_dir,
+            data_state=data_state,
+            logger=logger,
+            to_array=to_array,
+        )
```
Comment (Simsso) [inline src/data.py:393]: else branch with raise ValueError

File: src/predict.py [hunk 21]
```diff
@@ -34,20 +32,30 @@ def predict_from_torch(
...
     all_predictions = []
+    all_probabilities = []
 
     with torch.no_grad():
         for batch_features, _ in data_loader:
             batch_features = batch_features.to(device)
-            outputs = model(batch_features).squeeze()
+            outputs = model(batch_features)
 
             if return_probabilities:
```
Comment (Simsso) [inline src/predict.py:42]: These boolean parameters are smell. It's better to have one function returning logits (pre-softmax) and one which is called `predict_probas_from_torch` or so which calls the former and applies the softmax. As done in the new model classes you added.

File: src/predict.py [hunk 22]
```diff
@@ -72,4 +80,5 @@ def save_predictions_to_file(
             if probabilities is None:
                 f.write(f"{label}\n")
             else:
-                f.write(f"{label},{probabilities[idx]}\n")
+                probas_str = ",".join([str(p) for p in probabilities[idx]])
```
Comment (Simsso) [inline src/predict.py:83]: Can be written as `",".join(map(str, probabilities[idx]))` if you like that.

File: src/train.py [hunk 23]
```diff
@@ -217,10 +199,15 @@ def gridsearch_lgbm(
 
     logger.info("Performing a grid search for the best parameters...")
     grid = GridSearchCV(estimator=model, param_grid=config.model.param_grid, cv=5)
-    grid.fit(train_set.x, train_set.y)
+
+    try:
+        grid.fit(train_set.x, train_set.y)
+    except ValueError:  # if the data is not in the right shape, try squeezing it
+        grid.fit(train_set.x.squeeze(), train_set.y)
```
Comment (Simsso) [inline src/train.py:206]: Don't use try/except for normal, expected control flow. Add an auxiliary function which checks this and squeezes if needed.

File: src/utils.py [hunk 24]
```diff
@@ -193,3 +193,49 @@ def load_model(run_folder: str, model: Optional[nn.Module] = None) -> nn.Module:
...
+        Tuple containing the initialized model, optimizer, and loss criterion.
+    """
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+    model_class = config.model.model_type
+
+    if model_class in [
+        ModelType.LOGREG_TORCH,
+        ModelType.SIMPLE_CNN,
+        ModelType.SIMPLE_RNN,
+    ]:
```
Comment (Simsso) [inline src/utils.py:222]: If possible, use {} instead of []. (set instead of list)

**Review by Simsso — COMMENTED** [review]
Good job, Emily!

File: src/configs/config_scaffold.py [hunk 25]
```diff
@@ -8,6 +14,18 @@ class ModelFramework(Enum):
     LIGHTGBM = auto()  # for gradient boosting models
 
 
+class PyTorchLoss(Enum):
+    BINARY_CROSS_ENTROPY = nn.BCELoss  # commonly used for binary classification
+    CROSS_ENTROPY = nn.CrossEntropyLoss  # commonly used for multi-class classification
+    MSE = nn.MSELoss  # commonly used for regression
```
[the remark this answers is not in the document]
Comment (erobitschek) [reply src/configs/config_scaffold.py:20]: Do you think it adds unnecessary complexity?

File: src/configs/config_scaffold.py [hunk 26]
```diff
@@ -8,6 +14,18 @@ class ModelFramework(Enum):
...
+
+
+class PyTorchOptim(Enum):
+    ADAM = optim.Adam
+    SGD = optim.SGD
+    RMS_PROP = optim.RMSprop
```
[the remark this answers is not in the document]
Comment (erobitschek) [reply src/configs/config_scaffold.py:26]: hmm in what scenarios is "explicit control" useful or not?

File: src/configs/config_scaffold.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (erobitschek) [reply src/configs/config_scaffold.py:20]: took it out for now :) (and the optimizer enum)

File: src/configs/config_scaffold.py [hunk 27]
```diff
@@ -46,16 +74,46 @@ def __post_init__(self):
...
     name: str
     learning_rate: float
-    batch_size: int
+    batch_size: int = 32
     epochs: int = 500
+    model_type: ModelType = ModelType.LOGREG_SKLEARN
     framework: ModelFramework = ModelFramework.SKLEARN
+    loss_criterion: InitVar[
+        Optional[str]
+    ] = "CROSS_ENTROPY"  # default value for pytorch models
```
[the remark this answers is not in the document]
Comment (erobitschek) [reply src/configs/config_scaffold.py:83]: probably not necessary upon further review - removed

File: src/predict.py (hunk 22 above)
Comment (erobitschek) [reply src/predict.py:83]: I do like that :D

**Pull request merged**